### PR TITLE
BUILD-281:  add feature gate, OCM configuration bit, to enable build csi volume tech preview 

### PR DESCRIPTION
**Pull request by gabemontero**
/assign @adambkaplan 
/assign @bparees 
(given historical involvement in this effort, please feel free to unassign at your discretion)
/assign @deads2k 
(given recent reviews in this space as well as historical involvement in the shared/projected resource space; of course reassign at your discretion)

### CODE REVIEW

**Comment by gabemontero** [comment]
hold .... *THIS* story is also suppose to do the feature gate a la https://github.com/openshift/api/pull/982

**Comment by gabemontero** [comment]
unhold ... feature gate commit included 

PTAL

File: openshiftcontrolplane/v1/types.go [hunk 1]
```diff
@@ -296,6 +296,11 @@ type BuildDefaultsConfig struct {
 
 	// resources defines resource requirements to execute the build.
 	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+
+	// buildCSIVolumes defines whether the instances of DockerBuildStrategy or SourceBuildStrategy
+	// can employ CSI volumes with any BuildVolume they specify.  Currently, this is only
+	// supported on clusters where the TechPreview feature set is enabled.
+	BuildCSIVolumes bool `json:"buildCSIVolumes,omitempty"`
```
Comment (bparees) [inline openshiftcontrolplane/v1/types.go:303]: this seems odd to be nested under "Defaults" vs being somewhere in the top level BuildControllerConfig object.

Comment (gabemontero) [reply openshiftcontrolplane/v1/types.go:303]: I tend to agree .... but the input from the description in https://issues.redhat.com/browse/BUILD-281 -

Openshift-controller-manager-operator needs to read the feature gate, pass it on to the build controller.
The build controller has its own configuration "API" - this was a relic of the 3.x master configuration that is not exposed to admins in OCP 4.x: https://github.com/openshift/api/blob/master/openshiftcontrolplane/v1/types.go#L198-L207

@adambkaplan - perhaps a hiccup during planning, and we should instead update https://github.com/openshift/api/blob/master/config/v1/types_build.go#L47-L77 ?

if not, please clarify

Comment (bparees) [reply openshiftcontrolplane/v1/types.go:303]: did you get those links backwards?
https://github.com/openshift/api/blob/85977bee07221f012896dcc53b77f44da0be0c4e/openshiftcontrolplane/v1/types.go#L198-L207 is where i'd expect to put this (as a peer to BuildDefaults, not nested within BuildDefaults).

i'm not sure if this new field warrants a whole new struct within BuildControllerConfig, or if we can just put the bool field directly into that struct....i'm inclined towards the latter, since the struct is already named "BuildControllerConfig"

Comment (gabemontero) [reply openshiftcontrolplane/v1/types.go:303]: no that is what I was thinking, but I like your take better @bparees 

first things first, and the change that greatly simplifies things on the OCM-O side, is settling on config/v1/types_builds.go instead of openshiftcontrolplane/v1/types.go

hopefully @adambkaplan does not have any gotcha for us when he sees this thread

Comment (bparees) [reply openshiftcontrolplane/v1/types.go:303]: ok, i'm clearly forgetting how we ended up with both:

https://github.com/openshift/api/blob/85977bee07221f012896dcc53b77f44da0be0c4e/config/v1/types_build.go#L19

and

https://github.com/openshift/api/blob/85977bee07221f012896dcc53b77f44da0be0c4e/openshiftcontrolplane/v1/types.go#L198

I suspect you're going to need to change both.  One is presumably the config resource that admins edit/stored in etcd, and the other is used internally to write/hold the config for the controller.

but my point was independent of that and was only that the new field should be a peer to the BuildDefaults field, not a member of its struct.

Comment (gabemontero) [reply openshiftcontrolplane/v1/types.go]: ok @bparees @adambkaplan 

1) field moved up a level
2) analogous fields in both config/v1 and openshiftcontrolplane/v1

PTAL

File: config/v1/types_build.go [hunk 2]
```diff
@@ -42,6 +42,11 @@ type BuildSpec struct {
 	// BuildOverrides controls override settings for builds
 	// +optional
 	BuildOverrides BuildOverrides `json:"buildOverrides"`
+
+	// BuildCSIVolumes defines whether the instances of DockerBuildStrategy or SourceBuildStrategy
+	// can employ CSI volumes with any BuildVolume they specify.  Currently, this is only
+	// supported on clusters where the TechPreview feature set is enabled.
+	BuildCSIVolumes bool `json:"buildCSIVolumes,omitempty"`
```
Comment (bparees) [inline config/v1/types_build.go:49]: @deads2k do we typically prefix things like this with `Enable` or do we explicitly avoid that?

@gabemontero not sure what happened w/ the formatting here, but can you get the whitespace lined up in this struct?  I'm surprised this isn't failing gofmt validation.

File: openshiftcontrolplane/v1/types.go [hunk 3]
```diff
@@ -204,6 +204,11 @@ type BuildControllerConfig struct {
 	// additionalTrustedCA is a path to a pem bundle file containing additional CAs that
 	// should be trusted for image pushes and pulls during builds.
 	AdditionalTrustedCA string `json:"additionalTrustedCA"`
+
+	// buildCSIVolumes defines whether the instances of DockerBuildStrategy or SourceBuildStrategy
+	// can employ CSI volumes with any BuildVolume they specify.  Currently, this is only
+	// supported on clusters where the TechPreview feature set is enabled.
+	BuildCSIVolumes bool `json:"buildCSIVolumes,omitempty"`
```
Comment (bparees) [inline openshiftcontrolplane/v1/types.go:211]: same alignment issue on this struct.  And same question to @deads2k about whether we'd want to prefix with Enable

**Comment by bparees** [comment]
> analogous fields in both config/v1 and openshiftcontrolplane/v1

@gabemontero does that mean you confirmed what i said earlier (that we need it in both places because one is the user facing config api, and the other is the internal config struct for the controller process?)

File: config/v1/types_build.go (hunk 2 above)
Comment (gabemontero) [reply config/v1/types_build.go:49]: just ran `gofmt -w ./config ./pkg` and `gofmt -w -l ./config ./pkg`it produced no changes @bparees 

also examined tab vs. spaces in the preceding lines with the new ones and they are consistent

not sure why git's diff is complaining

Comment (gabemontero) [reply config/v1/types_build.go:49]: also, I see no difference when I look at the code diffs from the PR in my browser @bparees

Comment (bparees) [reply config/v1/types_build.go:49]: bizarre, this is what i see in the browser:
image

but yeah, doesn't seem to be reflected in the file itself.

Comment (gabemontero) [reply config/v1/types_build.go:49]: on the naming convention, fwiw, I did look at the https://docs.openshift.com/container-platform/4.8/rest_api/understanding-compatibility-guidelines.html link you provided on the email list recently, and did not see anything that fit with what you are asking about

**Comment by gabemontero** [comment]
> > analogous fields in both config/v1 and openshiftcontrolplane/v1
> 
> @gabemontero does that mean you confirmed what i said earlier (that we need it in both places because one is the user facing config api, and the other is the internal config struct for the controller process?)

not confirmed @bparees  .... placeholder / anticipating the answer.  Still waiting on @adambkaplan to catch up to this, as he was the one who cited updating openshiftcontrollplane

**Comment by gabemontero** [comment]
> > > analogous fields in both config/v1 and openshiftcontrolplane/v1
> > 
> > 
> > @gabemontero does that mean you confirmed what i said earlier (that we need it in both places because one is the user facing config api, and the other is the internal config struct for the controller process?)
> 
> not confirmed @bparees .... placeholder / anticipating the answer. Still waiting on @adambkaplan to catch up to this, as he was the one who cited updating openshiftcontrollplane

actually @bparees @adambkaplan .... if we were to take the discussion thread  with @deads2k at https://coreos.slack.com/archives/C014MHHKUSF/p1631216153043900 even further, and mimic the use of library-go for features gates like the kubeapiserver, then we would use the existing https://github.com/openshift/api/blob/master/openshiftcontrolplane/v1/types.go#L48, which is analogous to https://github.com/openshift/api/blob/master/kubecontrolplane/v1/types.go#L64, which is what is currently used by https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L134

the only fly in the ointment there is those TODO comments above the lines in those first 2 links

**Comment by gabemontero** [comment]
OK @bparees @adambkaplan @deads2k 

this revamped version attempts to capture/implement guidance from various slack conversations with each of you in some fashion, along with some resolution of items previously discussed here

Key points:
- leverage library-go's feature gate observe function
- only need to update the OCM config; hard to reconcile the feature gate observe functionality in library-go with the observe build functionality in OCM-0; plus really only need 1 on/off switch, which goes away once use of csi volumes for builds goes GA
- a bit of back and forth with @deads2k on the differing config approaches between apiserver(s) and OCM, but if I read him correctly, in the end, and with the awareness that feature gates are eventually going away, and with the awareness that our coupling with the cluster storage operator for csi shared resources, and CSO's extensive use of the tech preview feature gate, I've added a "feature gate centric" field to the OCM  config

Also, I have WIP PRs for illustrating the end to end flow between library-go, OCM-O, and OCM, and the minimal amount of code delta required, to achieve are means.

Please consider 
https://github.com/openshift/library-go/pull/1201
https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/227
https://github.com/openshift/openshift-controller-manager/pull/199

which are all based on the current commit level of this PR at the time of this comment

thanks

**Review by adambkaplan — COMMENTED** [review]
First, necessary background:

- https://github.com/openshift/api/blob/master/config/v1/types_build.go#L19-L45 is the build controller configuration that is stored in etcd and configurable by the admin.
- This configuration (and others) gets rendered into the `spec.observedConfig` field of the operator OpenshiftControllerManager object: https://github.com/openshift/api/blob/master/operator/v1/types_openshiftcontrollermanager.go#L15-L24.
- The schema for OpenshiftControllerManager's observed config is the old 3.x controller manager configuration, which can be expressed as JSON/YAML: https://github.com/openshift/api/blob/master/openshiftcontrolplane/v1/types.go#L164.
- Ultimately openshift-controller-manager-operator merges the default, observed, and unsupported overrides into the OpenshiftControllerManager JSON, and mounts this as a ConfigMap: https://github.com/openshift/cluster-openshift-controller-manager-operator/blob/master/pkg/operator/sync_openshiftcontrollermanager_v311_00.go#L183-L186

So the net is that the API in `/config` is stored in etcd, the API in `/openshiftcontrolplane` is not.

In this approach, the library-go function used will parse out the feature gates and render an array of strings with the format `<FeatureGate>=<true|false>`. The `FeatureGates` field therefore should just be `[]string` if we move forward with this approach. That said, we would then need extra logic in openshift-controller-manager to parse the feature gates, which to me feels like a bit of overkill.

When I scoped the story out, my thought is we could do the following:

- Add a field to BuildControllerConfig indicating if CSI volumes are enabled: https://github.com/openshift/api/blob/master/openshiftcontrolplane/v1/types.go#L198
- Update our existing build config observation function to read the cluster feature gates, check if `BuildCSIVolumes` is enabled, and if so, set the appropriate value. See https://github.com/openshift/cluster-openshift-controller-manager-operator/blob/master/pkg/operator/configobservation/builds/observe_builds.go#L18

This would avoid the "kube feature gates" -> "controller manager config" translation, which IMO feels brittle. The downside is that we would more or less need to replicate the logic in the cluster-storage-operator - see https://github.com/openshift/cluster-storage-operator/blob/2e76e02cb8479784a582e57a587be8fd5d75cae2/pkg/operator/csidriveroperator/driverstarter.go#L268-L292

File: openshiftcontrolplane/v1/types.go [hunk 4]
```diff
@@ -189,6 +189,11 @@ type OpenShiftControllerManagerConfig struct {
 	Ingress            IngressControllerConfig          `json:"ingress"`
 	ImageImport        ImageImportControllerConfig      `json:"imageImport"`
 	SecurityAllocator  SecurityAllocator                `json:"securityAllocator"`
+
+	// featureGates captures the current set of OpenShift feature gates that are of interest to
+	// components of the OpenShift Controller Manager.  NOTE: this field will possibly be deprecated
+	// in a future OpenShift release if the feature gate mechanism in OpenShift is removed.
+	FeatureGates map[string][]string `json:"featureGates"`
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply openshiftcontrolplane/v1/types.go:196]: My testing of things on Friday as I recall came to the conclusion that anything set to false did not get included in what ultimately bubbled up.  i.e. it got filtered at the library-go level.

I'll double check if we move forward down this path.  But if I'm incorrect, then yes, ocm-o/ocm would need to account for it.

In the interest of code reuse and expediency, I at least am OK with it.

**Comment by gabemontero** [comment]
Thanks for catching up to this after your kubecon foray @adambkaplan  ... some opnion/responses embedded below.

> First, necessary background:
> 
>     * https://github.com/openshift/api/blob/master/config/v1/types_build.go#L19-L45 is the build controller configuration that is stored in etcd and configurable by the admin.
> 
>     * This configuration (and others) gets rendered into the `spec.observedConfig` field of the operator OpenshiftControllerManager object: https://github.com/openshift/api/blob/master/operator/v1/types_openshiftcontrollermanager.go#L15-L24.
> 
>     * The schema for OpenshiftControllerManager's observed config is the old 3.x controller manager configuration, which can be expressed as JSON/YAML: https://github.com/openshift/api/blob/master/openshiftcontrolplane/v1/types.go#L164.
> 
>     * Ultimately openshift-controller-manager-operator merges the default, observed, and unsupported overrides into the OpenshiftControllerManager JSON, and mounts this as a ConfigMap: https://github.com/openshift/cluster-openshift-controller-manager-operator/blob/master/pkg/operator/sync_openshiftcontrollermanager_v311_00.go#L183-L186
> 
> 
> So the net is that the API in `/config` is stored in etcd, the API in `/openshiftcontrolplane` is not.
> 
> In this approach, the library-go function used will parse out the feature gates and render an array of strings with the format `<FeatureGate>=<true|false>`. The `FeatureGates` field therefore should just be `[]string` if we move forward with this approach. That said, we would then need extra logic in openshift-controller-manager to parse the feature gates, which to me feels like a bit of overkill.
> 
> When I scoped the story out, my thought is we could do the following:
> 
>     * Add a field to BuildControllerConfig indicating if CSI volumes are enabled: https://github.com/openshift/api/blob/master/openshiftcontrolplane/v1/types.go#L198
> 
>     * Update our existing build config observation function to read the cluster feature gates, check if `BuildCSIVolumes` is enabled, and if so, set the appropriate value. See https://github.com/openshift/cluster-openshift-controller-manager-operator/blob/master/pkg/operator/configobservation/builds/observe_builds.go#L18
> 

OK @adambkaplan yes this gets to the crux of the questions @bparees and I had at the beginning of this PR.  The intent you outlined above ^^ now does not precisely with what came out of planning for BUILD-281 and what I currently see in the description there.  But no big deal wrt that specifically.  I think the crux of the decisions we need to make here is what you accurately summarized below.

> 
> This would avoid the "kube feature gates" -> "controller manager config" translation, which IMO feels brittle. The downside is that we would more or less need to replicate the logic in the cluster-storage-operator - see https://github.com/openshift/cluster-storage-operator/blob/2e76e02cb8479784a582e57a587be8fd5d75cae2/pkg/operator/csidriveroperator/driverstarter.go#L268-L292

Totally agree with the characterization of things. 

My current opinions are:
- tweaking library-go for this would need to take into account working on top of the existing support provided for the flag driven api server consumption.
- feature gates are going away per @deads2k  .... so I think any time spent in that space should be minimized
- feature gating build support for csi volumes is a temporary thing, independent of the fate of feature gates ... ultimately, we want this to be support that just exists like any other.  Developers can choose to not use it.  Administrators can prevent use of CSI volumes for builds at a per namespace level by simply not granting the `builder` SA permissions to access the `shares`

Bottom line, I think we can manage any "brittle" translation in the short term while dealing with a tech preview item, and just jettison the whole thing when we graduate to GA.

File: openshiftcontrolplane/v1/types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gabemontero) [reply openshiftcontrolplane/v1/types.go:196]: I'll pull your suggestion in manually @adambkaplan since I'll need to run `make update` anyway to get the verifications to pass ... save on some CI churn

thanks

**Comment by adambkaplan** [comment]
> * tweaking library-go for this would need to take into account working on top of the existing support provided for the flag driven api server consumption.

I think this PR will help: https://github.com/openshift/library-go/pull/1208

File: openshiftcontrolplane/v1/types.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gabemontero) [reply openshiftcontrolplane/v1/types.go]: godoc update pushed @adambkaplan

**Comment by adambkaplan** [comment]
> Bottom line, I think we can manage any "brittle" translation in the short term while dealing with a tech preview item, and just jettison the whole thing when we graduate to GA.

This is fair. I'm now sold on using `FeatureGates` in the OCM configuration spec. However I think the type needs to be changed to `[]string` for us to use it properly.

**Comment by adambkaplan** [comment]
Confirmed - the observe features function in library-go will write a `[]string` to the raw JSON that is embedded in the ConfigMap:

```
data:
  config.yaml: |
    {"apiVersion":"openshiftcontrolplane.config.openshift.io/v1","build":{"buildDefaults":{"resources":{}},"imageTemplateFormat":{"format":"registry.build01.ci.openshift.org/ci-ln-skv518b/stable@sha256:7e9d76442b6ae3b7c19f7163b65ae0d5334fbbac42dfb50b62f6afc63f239634"}},"deployer":{"imageTemplateFormat":{"format":"registry.build01.ci.openshift.org/ci-ln-skv518b/stable@sha256:b356a5173fd2a2f4cce8b1bd7640cb94e5375a778409b72d12434196c6936eb3"}},"dockerPullSecret":{"internalRegistryHostname":"image-registry.openshift-image-registry.svc:5000"},"featureGates":["BuildCSIVolumes=true"],"ingress":{"ingressIPNetworkCIDR":""},"kind":"OpenShiftControllerManagerConfig"}
...
```

**Review by adambkaplan — APPROVED** [review]
/lgtm

@deads2k build team is in agreement with this approach.
